[Merged by Bors] - feat: port RepresentationTheory.Rep

101damnations opened the pull request:
---

Trying again to port this now the mathlib3 PR [#19170](https://github.com/leanprover-community/mathlib/pull/19170) is merged. The first port attempt was #4761.
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

jcommelin commented on Mathlib/RepresentationTheory/Rep.lean:
```diff
@@ -59,88 +60,107 @@ instance : CoeSort (Rep k G) (Type u) :=
 #align Rep.monoidal_category.braiding_hom_apply Rep.MonoidalCategory.braiding_hom_apply
 
-@[simp]
+/- Porting note: same note as above. -/
```
```suggestion
/- Porting note: linter informs me the LHS simplifies to... itself.
But fwiw I also can't get simp to use this lemma. -/
```
It's better to repeat the note, in case the other one changes or is removed, but this one isn't.

101damnations replied:
thanks, done!

jcommelin commented on Mathlib/RepresentationTheory/Rep.lean:
```diff
@@ -159,108 +180,140 @@ variable {k G}
+  exact LinearMap.ext fun x => LinearEquiv.symm_apply_apply _ _-/
 @[simp]
 theorem linearization_μ_inv_hom (X Y : Action (Type u) (MonCat.of G)) :
     (inv ((linearization k G).μ X Y)).hom = (finsuppTensorFinsupp' k X.V Y.V).symm.toLinearMap := by
```
```suggestion
@[simp]
theorem linearization_μ_inv_hom (X Y : Action (Type u) (MonCat.of G)) :
    (inv ((linearization k G).μ X Y)).hom = (finsuppTensorFinsupp' k X.V Y.V).symm.toLinearMap := by
-- Porting note: broken proof was
/-simp_rw [← Action.forget_map, Functor.map_inv, Action.forget_map, linearization_μ_hom]
  apply IsIso.inv_eq_of_hom_inv_id _
  exact LinearMap.ext fun x => LinearEquiv.symm_apply_apply _ _-/
```

101damnations replied:
thanks, done!

jcommelin commented on Mathlib/RepresentationTheory/Rep.lean:
```diff
@@ -326,74 +393,83 @@ protected theorem ihom_obj_ρ_apply {A B : Rep k G} (g : G) (x : A →ₗ[k] B)
 
-variable (A B C)
-
 /- ./././Mathport/Syntax/Translate/Expr.lean:177:8: unsupported: ambiguous notation -/
```
```suggestion
```
Lines like this can be deleted.

jcommelin reviewed (commented):
Thanks for leaving detailed porting notes. Would you please move them right above the point where the proof changed, instead of above the decl?

101damnations commented:
Accidentally removed a `nolint simpNF` in the last commit - I'm going to modify some `@[simps]` attributes in other files as [suggested](https://leanprover.zulipchat.com/#narrow/stream/144837-PR-reviews/topic/!4.235041.20RepresentationTheory.2ERep/near/366297267) by Scott, so that `simp` stops unfolding types and meaning I have to add `nolint simpNF`s.

kbuzzard reviewed (approved):
This looks fine to me. I've left some of my usual pernickety comments.

101damnations commented on Mathlib/RepresentationTheory/Rep.lean:
```diff
@@ -0,0 +1,667 @@
+import Mathlib.Algebra.Category.ModuleCat.Colimits
+import Mathlib.Algebra.Category.ModuleCat.Monoidal.Closed
+import Mathlib.Algebra.Category.ModuleCat.Adjunctions
+import Mathlib.CategoryTheory.Closed.FunctorCategory
```
I could only remove one!

jcommelin commented on Mathlib/RepresentationTheory/Rep.lean:
```diff
@@ -0,0 +1,667 @@
+      adj := Adjunction.mkOfHomEquiv (
+      { homEquiv := Rep.homEquiv A
+        homEquiv_naturality_left_symm := fun _ _ => Action.Hom.ext _ _
+          (TensorProduct.ext' fun _ _ => rfl)
```
Do you mean an `ext` lemma?
I think in Lean 3 we didn't have it as an ext lemma globally, but enabled it locally on occasion.

101damnations replied:
I tried enabling it locally with high priority, but Lean still won't use it. (I can't find it enabled locally anywhere in mathlib3 either).

101damnations commented on Mathlib/RepresentationTheory/Rep.lean:
```diff
@@ -0,0 +1,676 @@
+  Action.instPreservesColimitsForget.{u} _ _
+
+/- Porting note: linter informs me the LHS simplifies to... itself.
+But fwiw I also can't get simp to use this lemma. -/
```
Scott is investigating the type-simplifying `@[simps]` lemmas that fire here and produce that diff. I tried removing them all, and `simp` is then happy to use my `nolint` lemmas (without me telling it to). But the `@[simps]` lemmas I removed are sometimes useful, and taking them away breaks stuff. 

As you say, though, I don't actually need my `nolint` lemmas to be `simp` lemmas, so I'm just gonna remove the `simp` attributes for now.

101damnations commented:
I've removed 2 of the `protected`s I added in [the mathlib3 PR](https://github.com/leanprover-community/mathlib/pull/19170) because I don't think they have a good reason to be there. Not sure if this is worth a porting note or a backport.

101damnations commented on Mathlib/Algebra/Category/ModuleCat/Limits.lean:
```diff
@@ -138,6 +138,9 @@ instance hasLimits : HasLimits (ModuleCat.{w} R) :=
 #align Module.has_limits ModuleCat.hasLimits
 
+instance (priority := high) hasLimits' : HasLimits (ModuleCat.{u} R) :=
+  ModuleCat.hasLimitsOfSize.{u, u, u}
```
It's something to do with universes; see the instance above it:
```
instance hasLimits : HasLimits (ModuleCat.{w} R) :=
  ModuleCat.hasLimitsOfSize.{w, w, u}
#align Module.has_limits ModuleCat.hasLimits

instance (priority := high) hasLimits' : HasLimits (ModuleCat.{u} R) :=
  ModuleCat.hasLimitsOfSize.{u, u, u}
```

@joelriou added it to fix the error you worked on [here](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/!4.234761.20universe.20issues). I don't really understand it at the moment :)

ChrisHughes24 commented:
bors merge
